[Doc] Fix UTF-8 encoding issues in documentation generation on Windows

alhridoy opened the pull request:
## Summary

Fixes #24360: Cannot build documentation locally without adding encoding to generate examples

This PR resolves Windows-specific encoding issues when building documentation locally by explicitly specifying UTF-8 encoding for file operations in documentation generation scripts.

## Problem

On Windows, Python's default file encoding is often `cp1252` (Windows-1252), which cannot handle Unicode characters like `\uff5c` (fullwidth vertical bar `｜`). This caused:

- **UnicodeEncodeError** when processing files with Unicode content
- **Documentation build failures** on Windows systems
- **Inability for Windows contributors** to build docs locally

### Error Examples
```
UnicodeEncodeError: 'charmap' codec can't encode character '\uff5c' in position 61: character maps to <undefined>
```

## Solution

### **Core Changes**
- **`docs/mkdocs/hooks/generate_examples.py`**: Added `encoding='utf-8'` to file operations
- **`docs/mkdocs/hooks/generate_argparse.py`**: Added `encoding='utf-8'` to file operations
- **Comprehensive testing**: Added Unicode handling tests
- **Documentation**: Added Windows development guide

### **Files Modified**
```python
# Before (problematic)
with open(self.main_file) as f:
    first_line = f.readline().strip()

with open(doc_path, "w+") as f:
    f.write(example.generate())

# After (fixed)
with open(self.main_file, encoding="utf-8") as f:
    first_line = f.readline().strip()

with open(doc_path, "w+", encoding="utf-8") as f:
    f.write(example.generate())
```

Commit:
feat: Add sliding window support to FlexAttention backend

Fixes #24358: FlexAttention does not support sliding window yet

This commit implements sliding window attention support for the FlexAttention
backend, enabling models like GptOss to work with FlexAttention when
FlashAttention v2 is unavailable (e.g., on GPUs with compute capability < 8.0).

Key changes:
- Add sliding_window_causal_mask_mod() function for sliding window masks
- Update FlexAttentionImpl to handle sliding window configuration
- Enhance FlexAttentionMetadata with sliding_window field
- Update FlexAttentionMetadataBuilder to extract sliding window from config
- Add comprehensive unit and integration tests

The implementation maintains backward compatibility and has no performance
impact on models without sliding window.

Tested with:
- Unit tests for mask functions and metadata handling
- Integration tests with FlexAttentionImpl
- Edge cases and error conditions
- Backward compatibility verification

Before this change:
- Models with sliding window would fail with NotImplementedError
- Users had to manually force different attention backends
- GptOss models couldn't use FlexAttention on older GPUs

After this change:
- FlexAttention supports sliding window attention seamlessly
- GptOss models work out of the box with FlexAttention
- Maintains full backward compatibility for models without sliding window

Signed-off-by: alekramelaheehridoy <aliqramalaheehridoy@gmail.com>

hmellor reviewed (changes requested):
Thank you for the fix which makes the docs build more flexible!

However, I don't think we need the doc or the tests.

vLLM is not supported on Windows (only via WSL), so development instructions for Windows itself might give the wrong idea. 

As for the tests, a comment next to the file opening lines saying `# Specify encoding for building on Windows` is sufficient.

alhridoy commented:
@hmellor Thank you for the feedback! You're absolutely right - I've removed the Windows documentation
   and comprehensive tests.

  The PR now contains only:
  - UTF-8 encoding fixes in the two documentation generation files
  - Simple comments explaining the encoding is for Windows builds

hmellor reviewed (approved):
Thanks for updating!

hmellor commented:
Please fix DCO then I'll be able to merge

alhridoy committed:
fix: Add UTF-8 encoding to documentation generation scripts

Fixes #24360: Cannot build documentation locally without adding encoding

This commit resolves Windows-specific encoding issues when building
documentation locally by explicitly specifying UTF-8 encoding for
file operations in documentation generation scripts.

Changes:
- Add encoding='utf-8' to file operations in generate_examples.py
- Add encoding='utf-8' to file operations in generate_argparse.py
- Add comprehensive tests for Unicode handling
- Add Windows development guide with encoding troubleshooting

Root Cause:
Windows uses cp1252 (Windows-1252) as default encoding, which cannot
handle Unicode characters like \uff5c (fullwidth vertical bar ｜).
This caused UnicodeEncodeError when processing files with Unicode content.

Solution:
Explicitly specify UTF-8 encoding for all file read/write operations
in documentation generation hooks, ensuring consistent behavior across
all platforms.

Testing:
- Added unit tests for Unicode character handling
- Tested with problematic characters from the original error
- Verified backward compatibility with ASCII-only content
- Added edge case testing for empty files and missing titles

Benefits:
- Enables local documentation building on Windows
- Improves cross-platform development experience
- Prevents similar encoding issues in the future
- Maintains full backward compatibility

Before this change:
- Documentation build failed on Windows with UnicodeEncodeError
- Contributors on Windows couldn't build docs locally
- Unicode characters in examples caused build failures

After this change:
- Documentation builds successfully on Windows
- Full Unicode support in documentation content
- Consistent behavior across all platforms

Signed-off-by: alekramelaheehridoy <alekramelaheehridoy@gmail.com>
Signed-off-by: alekramelaheehridoy <aliqramalaheehridoy@gmail.com>

alhridoy committed:
fix: Add UTF-8 encoding to documentation generation for Windows

Specify encoding explicitly in file operations to prevent UnicodeEncodeError
when building docs on Windows systems that default to cp1252 encoding.

Fixes #24360

Signed-off-by: alekramelaheehridoy <alekramelaheehridoy@gmail.com>
Signed-off-by: alekramelaheehridoy <aliqramalaheehridoy@gmail.com>

alhridoy committed:
fix: Remove unrelated FlexAttention changes from documentation PR

This commit removes FlexAttention sliding window support changes that were
incorrectly included in the documentation encoding fix PR. These changes
belong in a separate PR to maintain clean commit history and proper
separation of concerns.

Removed:
- FlexAttention sliding window implementation
- Related test files for sliding window functionality
- Unrelated changes to flex_attention.py

This PR should now focus solely on fixing UTF-8 encoding issues in
documentation generation scripts as originally intended.

Signed-off-by: alekramelaheehridoy <alekramelaheehridoy@gmail.com>
Signed-off-by: alekramelaheehridoy <aliqramalaheehridoy@gmail.com>

alhridoy commented:
@hmellor DCO is now fixed! All commits have been properly signed off with the correct email address. Thanks again!
